### DISCUSSION TITLE
Record the pitfalls of printing PDF files through custom labels.

### DISCUSSION THREAD

**Discussion by zly20129** [inventree/InvenTree#7736] [Show and tell]
- English version:

Title: Record the pitfalls of printing PDF files through custom labels.
[Related problem demonstration](https://github.com/inventree/InvenTree/issues/7735)
Tags can be customized through HTML files, such as the official part QR code and text printing template provided below:
```
{% extends "label/label_base.html" %}

{% load l10n i18n barcode %}

{% block style %}

.qr {
    position: absolute;
    left: 0mm;
    top: 0mm;
    {% localize off %}
    height: {{ height }}mm;
    width: {{ height }}mm;
    {% endlocalize %}
}

.part {
    font-family: Arial, Helvetica, sans-serif;
    display: flex;
    position: absolute;
    {% localize off %}
    left: {{ height }}mm;
    {% endlocalize %}
    top: 2mm;
}

{% endblock style %}

{% block content %}

<img class='qr' alt="{% trans 'QR Code' %}" src='{% qrcode qr_data %}'>

<div class='part'>
    {{ part.full_name }}
</div>

{% endblock content %}
```

The text part is defined by the `.part` segment, where the first item: `font-family` defines the set of fonts that can be used.
But the default of these three fonts are purely English fonts, does not support the Chinese printout, which leads to the printing of labels when the Chinese content is blank, while the English content can be displayed, such as 【0805电阻】, the actual output to the PDF only 【0805】.
The reason for this problem is that the official Docker container does not have Chinese font support installed by default. Using `docker exec inventree-server fc-list :lang=zh` you can see that the output is empty, which means that none of the Chinese fonts are installed in the container.
Since there is no font support, this results in no output - IMO, a solution idea can be constructed from this.
Solution (system is Ubuntu 20.04 Server):
1. First, install the Chinese font support `sudo apt-get install ttf-wqy-microhei` and `sudo apt-get install ttf-wqy-zenhei` on your operating system, and use these two commands to install Chinese open source fonts;
2. Use the command `fc-list :lang:zh` to check if the installation is successful, if the following result appears, the installation is successful;
```
/usr/share/fonts/truetype/dejavu/DejaVuSerif-Bold.ttf: DejaVu Serif:style=Bold
/usr/share/fonts/truetype/dejavu/DejaVuSansMono.ttf: DejaVu Sans Mono:style=Book
/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf: DejaVu Sans:style=Book
/usr/share/fonts/truetype/wqy/wqy-microhei.ttc: WenQuanYi Micro Hei,文泉驛微米黑,文泉驿微米黑:style=Regular
/usr/share/fonts/truetype/wqy/wqy-zenhei.ttc: WenQuanYi Zen Hei,文泉驛正黑,文泉驿正黑:style=Regular
/usr/share/fonts/truetype/wqy/wqy-zenhei.ttc: WenQuanYi Zen Hei Sharp,文泉驛點陣正黑,文泉驿点阵正黑:style=Regular
/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf: DejaVu Sans:style=Bold
/usr/share/fonts/truetype/dejavu/DejaVuSansMono-Bold.ttf: DejaVu Sans Mono:style=Bold
/usr/share/fonts/truetype/wqy/wqy-zenhei.ttc: WenQuanYi Zen Hei Mono,文泉驛等寬正黑,文泉驿等宽正黑:style=Regular
/usr/share/fonts/truetype/wqy/wqy-microhei.ttc: WenQuanYi Micro Hei Mono,文泉驛等寬微米黑,文泉驿等宽微米黑:style=Regular
/usr/share/fonts/truetype/dejavu/DejaVuSerif.ttf: DejaVu Serif:style=Book
```

3. In the output of the previous step, `/usr/share/fonts/truetype/wqy/wqy-microhei.ttc' and `/usr/share/fonts/truetype/wqy/wqy-zenhei.ttc' are the paths to the fonts that were installed by the two commands executed in step 1. Save this path temporarily;
4. Use `docker cp /usr/share/fonts/truetype/wqy/wqy-zenhei.ttc inventree-server:/usr/share/fonts/wqy-zenhei.ttc` and `docker cp /usr/share/fonts/wqy/microhei.ttc inventree-server:/usr/share/fonts/wqy-microhei.ttc`. fonts/truetype/wqy/wqy-microhei.ttc inventree-server:/usr/share/fonts/wqy-microhei.ttc` to copy each of the two fonts installed to your system into the containers;
5. Use `docker exec inventree-server fc-cache -fv /usr/share/fonts/wqy-zenhei.ttc` and `docker exec inventree-server fc-cache -fv /usr/share/fonts/ wqy-microhei.ttc` to install the two copied fonts into the server container environment;
6. In the `font-family:` paragraph of the `.part` of the custom tag HTML file, add `wqy-zenhei` or `wqy-microhei`, which is the font to be used in the Chinese language, or, of course, you can add all of the two fonts in descending order of priority from front to back, with the formatting without any quotation marks but separated by commas. without any quotes, but separated by commas;
P.S. Personally, I prefer the wqy-zenhei font.
7. Will you write a good custom label file uploaded to the server, and then test the printout, the rest of the format and other related custom content, please refer to the official documentation written.

Problems:
1. In the official update process, there is a container pull command operation, which will re-download the relevant containers from the server, so it can be determined that the font import and installation operations need to be re-executed after pulling the update.
2. After changing the .env configuration file or other related configuration files, updating and restarting the docker environment will result in the loss of installed fonts, and you need to re-execute the font import and installation operations.
Note: This refers to the docker container updating environment variables and rebooting, not the normal system reboot of the server!

- 中文版本：

标题：记录通过自定义标签打印PDF文件输出的坑。
[相关问题演示](https://github.com/inventree/InvenTree/issues/7735)
标签可通过HTML文件实现自定义，例如官方提供的零件二维码+文本打印模板如下：
```
{% extends "label/label_base.html" %}

{% load l10n i18n barcode %}

{% block style %}

.qr {
    position: absolute;
    left: 0mm;
    top: 0mm;
    {% localize off %}
    height: {{ height }}mm;
    width: {{ height }}mm;
    {% endlocalize %}
}

.part {
    font-family: Arial, Helvetica, sans-serif;
    display: flex;
    position: absolute;
    {% localize off %}
    left: {{ height }}mm;
    {% endlocalize %}
    top: 2mm;
}

{% endblock style %}

{% block content %}

<img class='qr' alt="{% trans 'QR Code' %}" src='{% qrcode qr_data %}'>

<div class='part'>
    {{ part.full_name }}
</div>

{% endblock content %}
```

文本部分是由`.part`段定义的，这段里边第一项：`font-family`定义了可以使用到的一系列字体。
但是默认的这三个字体都是纯英文字体，不支持中文打印输出，这就导致了打印标签的时候中文内容是一片空白，而英文内容则可以显示，例如0805电阻，实际输出到PDF的只有0805。
出现这种问题的原因就是官方的Docker容器默认并没有安装中文字体支持，使用`docker exec inventree-server fc-list :lang=zh`可以看到输出为空，这说明容器中并没有安装任何一个中文字体。
因为没有字体支持，所以导致没有输出——综上所述，可以由此构建解决思路。
解决方案（系统为Ubuntu20.04 Server）：
1. 首先在操作系统中安装中文字体支持`sudo apt-get install ttf-wqy-microhei`和`sudo apt-get install ttf-wqy-zenhei`，使用这两条指令安装中文开源字体；
4. 使用指令`fc-list :lang:zh`查看是否安装成功，如果出现以下结果，说明安装成功；
```
/usr/share/fonts/truetype/dejavu/DejaVuSerif-Bold.ttf: DejaVu Serif:style=Bold
/usr/share/fonts/truetype/dejavu/DejaVuSansMono.ttf: DejaVu Sans Mono:style=Book
/usr/share/fonts/truetype/dejavu/DejaVuSans.ttf: DejaVu Sans:style=Book
/usr/share/fonts/truetype/wqy/wqy-microhei.ttc: WenQuanYi Micro Hei,文泉驛微米黑,文泉驿微米黑:style=Regular
/usr/share/fonts/truetype/wqy/wqy-zenhei.ttc: WenQuanYi Zen Hei,文泉驛正黑,文泉驿正黑:style=Regular
/usr/share/fonts/truetype/wqy/wqy-zenhei.ttc: WenQuanYi Zen Hei Sharp,文泉驛點陣正黑,文泉驿点阵正黑:style=Regular
/usr/share/fonts/truetype/dejavu/DejaVuSans-Bold.ttf: DejaVu Sans:style=Bold
/usr/share/fonts/truetype/dejavu/DejaVuSansMono-Bold.ttf: DejaVu Sans Mono:style=Bold
/usr/share/fonts/truetype/wqy/wqy-zenhei.ttc: WenQuanYi Zen Hei Mono,文泉驛等寬正黑,文泉驿等宽正黑:style=Regular
/usr/share/fonts/truetype/wqy/wqy-microhei.ttc: WenQuanYi Micro Hei Mono,文泉驛等寬微米黑,文泉驿等宽微米黑:style=Regular
/usr/share/fonts/truetype/dejavu/DejaVuSerif.ttf: DejaVu Serif:style=Book
```

5. 上一步的输出结果中`/usr/share/fonts/truetype/wqy/wqy-microhei.ttc`和`/usr/share/fonts/truetype/wqy/wqy-zenhei.ttc`分别是第1步中所执行的两条指令所安装字体的路径，暂存这个路径；
6. 使用`docker cp /usr/share/fonts/truetype/wqy/wqy-zenhei.ttc inventree-server:/usr/share/fonts/wqy-zenhei.ttc`和`docker cp /usr/share/fonts/truetype/wqy/wqy-microhei.ttc inventree-server:/usr/share/fonts/wqy-microhei.ttc`，把安装到系统的两个字体分别复制到容器内；
7. 使用`docker exec inventree-server  fc-cache -fv /usr/share/fonts/wqy-zenhei.ttc`和`docker exec inventree-server  fc-cache -fv /usr/share/fonts/wqy-microhei.ttc`，将复制过去的两个字体安装到服务器容器环境中；
8. 在自定义标签HTML文件中的`.part`的`font-family:`段中，加入`wqy-zenhei`或者`wqy-microhei`，加入哪入中文用的就是哪种字体，当然，也可以按照从前到后优先级下降的顺序将两个全加进去，格式不用加任何的引号，但要加逗号分隔；
附：个人更喜欢wqy-zenhei字体。
9. 将你写好的自定义标签文件上传到服务器，然后测试打印输出，剩下的格式等相关的自定义内容请参照官方文档编写。

存在的问题：
1. 官方给出的更新过程中，有一项容器拉取的指令操作，这个操作会从服务器重新下载相关的容器，可以确定在拉取更新后需要重新执行字体导入及安装操作。
2. 更改.env配置文件或其他相关配置文件后，更新并重启docker环境会导致安装的字体丢失，需要重新执行字体导入及安装操作。
注意：这里指的是docker容器更新环境变量并重启，而不是服务器的普通系统重启！

**Comment by SchrodingersGat** [comment]
Hi @zly20129 please keep any issues / discussions limited to English language on the github pages. While the project code is translated, these pages are not and the development team's common language is English.

**Reply by zly20129** [reply]
Okay, I will rewrite an English version of the instructions.

**Reply by zly20129** [reply]
I've written up the English version for your perusal.

I've kept a copy of the Chinese version on the back burner for those who need it, as this problem only affects users of double-width character languages, especially Chinese users.
